Doc/library/pathlib.rst

```diff
@@ -162,6 +162,15 @@ we also call *flavours*:
    to ``PurePosixPath('bar')``, which is wrong if ``foo`` is a symbolic link
    to another directory)
 
+   The *flavour* keyword-only argument is available only in user subclasses of
+   :class:`PurePath` and :class:`Path`. It specifies the implementation of
+   :mod:`os.path` to use for low-level path operations. It may be left unset
+   or set to :mod:`os.path` to use the current system's flavour, or set to
+   ``posixpath`` or ``ntpath`` to use POSIX or Windows path semantics.
```

Comment:
1: could posixpath and ntpath on this line use `mod` markup?
2: should the doc say that an arbitrary object can be passed (like in https://github.com/fsspec/universal_pathlib/issues/114) to implement virtual filesystems?
3: isn’t it confusing to note a parameter in the class constructor doc (line 111) when it’s not for that class but only subclasses?

Reply:
> could posixpath and ntpath on this line use `mod` markup?

I tried this initially, but it produces Sphinx errors because `ntpath` and `posixpath` are never declared with `.. module::`. I could adjust the `os.path` docs to include these directives, perhaps?

> should the doc say that an arbitrary object can be passed (like in [fsspec/universal_pathlib#114](https://github.com/fsspec/universal_pathlib/issues/114)) to implement virtual filesystems?

So far I haven't found a use case for anything other than `posixpath` or `ntpath`. Most virtual filesystems use POSIX-style paths, e.g. FTP, tar files.

If we allowed arbitary objects, we'd need to define the interface in the docs. It looks like this at the moment:

```
sep
altsep
join()
normcase()
splitroot()
splitdrive()
isabs()
samestat()
```

Of these, we should be able to eliminate [`join()`](https://github.com/python/cpython/pull/105484) and `splitdrive()`.

... but this stuff is very low-level IMHO, and probably not interesting to 99% of users.

> isn’t it confusing to note a parameter in the class constructor doc (line 111) when it’s not for that class but only subclasses? but wouldn’t it be better to keep flavour as a class attribute, defined one in subclasses and not required for each instanciation?

Arguably yes, and it would simplify the implementation too!

But it would mean that subclasses' flavours are fixed, and can't be determined at initialisation time. Does that matter? I'm not sure. But if we commit to making it a class attribute, it will be quite difficult to change in future.

Reply:
I've asked about point (3) here: https://discuss.python.org/t/make-pathlib-extensible/3428/124

Reply:
1: Ok, no need to do these changes in os.path docs.  I’ll review the wording again to make sure it’s super clear that the value is a module object.

2: Ah I understand, pathmod is only used for lexical operations not data access!  (I think I was misled by the note from upath that they would pass a custom module, plus memories of the accessor classes)

3: I don’t know of any need to pass a flavour/pathmod at init time.  Thanks to your review of third-party path libs and your discussions with them, if you find a need for that then the init param could stay, otherwise a class attribute is easy and clear and doesn’t cause problems.  And it would still be possible to make it an instance variable and init param later, as instance attributes win over class attributes!

Reply:
1: I've tweaked the wording to make it clearer that it's a module, hopefully

2: There actually _were_ a few places where `self.pathmod` was used for non-lexical operations. As the value was guaranteed to be `os.path` in `Path`, it had no practical effect. But for clarity's sake, I've changed them to use `os.path` directly instead.

3: Having surveyed the existing pathlib-y third party libraries, I can't see a good reason to make it a class attribute. So I've switched it to a class attribute in this PR!